src/api/udf/Histogram.h

```diff
@@ -0,0 +1,161 @@
+      histogram_{ bucketSize_, min_, max_ } {
+    }
+    virtual ~Aggregator() = default;
+
+    inline virtual void merge(InputType v) override {
+      histogram_.addValue(v);
+    }
+
+    inline virtual void mix(const nebula::surface::eval::Sketch& another) override {
+      auto right = static_cast<const Aggregator&>(another);
+      histogram_.merge(right.histogram_);
+    }
+
+    inline virtual NativeType finalize() override {
+      json_ = jsonfy();
+      return json_;
+    }
+
+    // serialize into a buffer
+    inline virtual size_t serialize(nebula::common::ExtendableSlice& slice, size_t offset) override {
+      auto jsonStr = jsonfy();
+      auto bin = slice.write(offset, jsonStr.size());
+      auto size = slice.write(offset + bin, jsonStr.c_str(), jsonStr.size());
+      return bin + size;
+    }
+
+    inline virtual size_t load(nebula::common::ExtendableSlice& slice, size_t offset) override {
+      auto size = slice.read<size_t>(offset);
+      auto jsonStr = std::string(slice.read(offset + SIZE_SIZE, size));
+      histogram_ = folly::Histogram<InputType>(bucketSize_, min_, max_);
+      rapidjson::Document document;
+      if (document.Parse(jsonStr.c_str()).HasParseError()) {
+        throw NException(fmt::format("Error parsing histogram json: {0}", jsonStr));
+      }
+      for (rapidjson::Value::ConstMemberIterator itr = document.MemberBegin(); itr != document.MemberEnd(); ++itr) {
+        int64_t idx = 0;
+        uint64_t bucketCount = 0;
+        InputType bucketSum = 0;
+        for (auto& v : itr->value.GetArray()) {
+          if (idx == 2) {
+            bucketCount = std::stoi(v.GetString());
+          } else if (idx == 3) {
+            bucketSum = static_cast<InputType>(std::stod(v.GetString()));
+          }
+          idx++;
+        }
+        histogram_.addRepeatedValue(bucketSum / bucketCount, bucketCount);
+      }
+
+      return size + SIZE_SIZE;
+    }
+
+    inline virtual bool fit(size_t) override {
+      return false;
+    }
+
+  private:
+    InputType min_;
+    InputType max_;
+    size_t bucketNum_;
+    InputType bucketSize_;
+    folly::Histogram<InputType> histogram_;
+    std::string json_;
+
+    std::string jsonfy() const noexcept {
+      // Save histogram in below json format:
+      // {"bucketIndex": ["minVal", "maxVal", "count", "sum"]}
+      rapidjson::StringBuffer buffer;
+      buffer.Clear();
+      rapidjson::Writer<rapidjson::StringBuffer> json(buffer);
+      json.StartObject();
+      for (size_t i = 0; i < histogram_.getNumBuckets(); ++i) {
+        if (histogram_.getBucketByIndex(i).count == 0) {
+          continue;
+        }
+        const auto& bucket = histogram_.getBucketByIndex(i);
+        auto index = std::to_string(i);
+        json.Key(index.data(), index.size());
+        json.StartArray();
+        std::ostringstream bucketMin;
+        bucketMin << histogram_.getBucketMin(i);
+        json.String(bucketMin.str().data(), bucketMin.str().size());
```

Comment:
instead of converting everything to string, we can always use int64 for non-floating InputType and double for floating InputType to write min, max, and sum, otherwise this may be too slow when scanning billions of rows.

We can always use json.int64 for count.

Reply:
always use double for sum is fine too.

Reply:
thanks Shawn, changed per your feedback, please take another look when you get time. thanks! 🙏